src/replit_river/common_session.py

```diff
@@ -29,107 +33,69 @@ async def __call__(
 class SessionState(enum.Enum):
     """The state a session can be in.
 
-    Can only transition from ACTIVE to CLOSING to CLOSED.
+    Valid transitions:
+    - PENDING -> {CONNECTING}
+    - CONNECTING -> {PENDING, ACTIVE, CLOSING}
+    - ACTIVE -> {PENDING, CONNECTING, CLOSING}
+    - CLOSING -> {CLOSED}
+    - CLOSED -> {}
     """
 
-    ACTIVE = 0
-    CLOSING = 1
-    CLOSED = 2
+    PENDING = 0
+    CONNECTING = 1
+    ACTIVE = 2
+    CLOSING = 3
+    CLOSED = 4
 
 
-async def setup_heartbeat(
-    session_id: str,
-    heartbeat_ms: float,
-    heartbeats_until_dead: int,
-    get_state: Callable[[], SessionState],
-    get_closing_grace_period: Callable[[], float | None],
-    close_websocket: Callable[[], Awaitable[None]],
-    send_message: SendMessage,
-    increment_and_get_heartbeat_misses: Callable[[], int],
+ConnectingStates = set([SessionState.PENDING, SessionState.CONNECTING])
+TerminalStates = set([SessionState.CLOSING, SessionState.CLOSED])
+
+
+async def buffered_message_sender(
+    connection_condition: asyncio.Condition,
+    message_enqueued: asyncio.Semaphore,
+    get_ws: Callable[[], WebSocketCommonProtocol | ClientConnection | None],
+    websocket_closed_callback: Callable[[], Coroutine[Any, Any, None]],
+    get_next_pending: Callable[[], TransportMessage | None],
+    commit: Callable[[TransportMessage], None],
 ) -> None:
-    logger.debug("Start heartbeat")
     while True:
-        await asyncio.sleep(heartbeat_ms / 1000)
-        state = get_state()
-        if state != SessionState.ACTIVE:
+        await message_enqueued.acquire()
+        while (ws := get_ws()) is None:
+            # Block until we have a handle
```

Comment:
what happens if we never reconnect?

Reply:
If we never reconnect, `close()` now sends an event to wake up these functions so we can get caught by

```python
        if not ws:
            logger.debug("_buffered_message_sender: ws is not connected, loop")
            continue
```

which `continues` back to the outer loop, which terminates due to

```python
    while our_task and not our_task.cancelling() and not our_task.cancelled():
```

... being false.

If we have a connection interruption, the outer loop will not be false and we'll go back into `await block_until_connected()`.